pkgs/by-name/ni/nixos-jobserver/nixos-jobserver.cpp

```diff
@@ -0,0 +1,463 @@
+    .open = jobserver_open,
+    .read = jobserver_read,
+    .write = jobserver_write,
+    .release = jobserver_release,
+    .setxattr = jobserver_setxattr,
+    .getxattr = jobserver_getxattr,
+    .listxattr = jobserver_listxattr,
+    .removexattr = jobserver_removexattr,
+    .ioctl = jobserver_ioctl,
+    .poll = jobserver_poll,
+};
+
+static void print_help(std::ostream& to)
+{
+    to << R"(usage: nixos-jobserver -t <tokens> <mount point>
+
+starts a job server with the given number of tokens at the specified mount
+point. must be started as root to allow other users access to the jobserver.
+
+arguments:
+
+  -h           print help text
+  -t <tokens>  set number of tokens available (uses number of CPUs when 0)
+  -u <user>    mount jobserver with owner <user>
+  -g <group>   mount jobserver with group <group>
+  -d           debug mode (log fuse requests)
+)";
+}
+
+int main(int argc, char* argv[])
+{
+    std::string mount, user, group;
+    std::vector<std::string> args{
+        argv[0],
+        "-odefault_permissions",
+    };
+
+    if (geteuid() == 0) {
+        args.push_back("-oallow_other");
+    }
+
+    for (;;) {
+        int opt = getopt(argc, argv, "hdu:g:t:");
+        if (opt == -1)
+            break;
+        switch (opt) {
+        case 'd':
+            args.push_back("-d");
+            break;
+        case 'u':
+            user = optarg;
+            break;
+        case 'g':
+            group = optarg;
+            break;
+        case 't': {
+            char* end;
+            auto res = strtoul(optarg, &end, 10);
+            if (errno || isspace(optarg[0]) || *end || res > std::numeric_limits<unsigned>::max()) {
+                std::cerr << "bad argument for -t" << std::endl;
+                return 1;
+            }
+            tokens = res;
+            break;
+        }
+        case 'h':
+            print_help(std::cout);
+            return 0;
+        default:
+            print_help(std::cerr);
+            return 1;
+        }
+    }
+
+    if (optind >= argc) {
+        std::cerr << "missing mount point" << std::endl;
+        return 1;
+    } else if (optind + 1 < argc) {
+        std::cerr << "extraneous arguments after mount point" << std::endl;
+        return 1;
+    } else {
+        mount = argv[optind];
+    }
+
+    if (tokens == 0) {
+        tokens = get_nprocs_conf();
+    }
+
+    if (!user.empty()) {
+        if (auto pw = getpwnam(user.c_str()); pw) {
+            jobserver_st.st_uid = pw->pw_uid;
+        } else if (uid_t uid; sscanf(user.c_str(), "%u%c", &uid, (char*) &uid) == 1) {
+            jobserver_st.st_uid = uid;
+        } else {
+            std::cerr << "invalid argument for -u" << std::endl;
+            return 1;
+        }
+    }
+
+    if (!group.empty()) {
+        if (auto gr = getgrnam(group.c_str()); gr) {
+            jobserver_st.st_gid = gr->gr_gid;
+        } else if (gid_t gid; sscanf(group.c_str(), "%u%c", &gid, (char*) &gid) == 1) {
```

Comment:
it feels mildly evil to do that rather than make a second scratch variable that doesn't do anything, since one could have "123a", probably? I'm not sure if it's an aliasing rule violation but it feels somewhat ehh.